File: backend/ml_engine/rule_engine.py
"""
rule_engine.py
--------------
Recommendation & Risk Scoring Engine adapted for the dictionary-based
knowledge base schema.
"""

import re
from typing import List, Dict, Any, Optional, Union

CURRENT_YEAR = 2026
DEFAULT_BUFFER_YEARS = 4

THREAT_BUFFER_MODIFIER = {
    "Low": -1,
    "Medium": 0,
    "High": 1,
    "Critical": 2,
}

RISK_GREEN = "🟢"
RISK_YELLOW = "🟡"
RISK_RED = "🔴"

RISK_RANK = {RISK_GREEN: 0, RISK_YELLOW: 1, RISK_RED: 2}


# ---------------------------------------------------------------------------
# HELPER FUNCTIONS
# ---------------------------------------------------------------------------

def _extract_deprecation_year(security_lifespan: str) -> Optional[int]:
    """Extract numeric year from security lifespan string descriptions."""
    if not security_lifespan:
        return None
    matches = re.findall(r'\b(20\d\d)\b', security_lifespan)
    if matches:
        return int(matches[0])
    return None


def _is_quantum_safe(entry: Dict[str, Any]) -> bool:
    """Assess whether the recommended algorithm or approach is post-quantum secure."""
    text = (
        f"{entry.get('pqcMigrationPath', '')} "
        f"{entry.get('rationale', '')} "
        f"{entry.get('recommendedAlgorithm', '')}"
    ).lower()

    if "no migration needed" in text or "not quantum-vulnerable" in text or "not applicable" in text:
        return True
    if "crystals" in entry.get("recommendedAlgorithm", "").lower():
        return True
    return False


def _normalize_key(key: str) -> str:
    """Normalize input strings for fuzzy key matching."""
    return re.sub(r'[^a-zA-Z0-9]', '', str(key)).lower()


def _resolve_use_case_entry(
    use_case: str,
    db: Union[Dict[str, Any], List[Dict[str, Any]]]
) -> Optional[Dict[str, Any]]:
    """Locate the target use case within the dictionary or list-based knowledge base."""
    if isinstance(db, list):
        for item in db:
            if item.get("useCase", "").lower() == use_case.lower() or item.get("use_case", "").lower() == use_case.lower():
                return item
        return None

    norm_target = _normalize_key(use_case)

    for k, v in db.items():
        if _normalize_key(k) == norm_target or _normalize_key(v.get("useCase", "")) == norm_target:
            return v

    synonyms = {
        "otaupdates": "Signatures",
        "ota": "Signatures",
        "firmwaresigning": "Signatures",
        "firmware": "Signatures",
        "v2x": "Communication",
        "v2xcommunication": "Communication",
        "can": "MAC_SecOC",
        "secoc": "MAC_SecOC",
        "storage": "DataEncryption",
        "encryption": "DataEncryption",
        "boot": "SecureBoot",
        "lifecycle": "KeyManagement",
        "provisioning": "HSMProvisioning",
    }

    mapped_key = synonyms.get(norm_target)
    if mapped_key and mapped_key in db:
        return db[mapped_key]

    return None


# ---------------------------------------------------------------------------
# RISK SCORING
# ---------------------------------------------------------------------------

def _compute_buffer_years(threat_level: str) -> int:
    modifier = THREAT_BUFFER_MODIFIER.get(threat_level, 0)
    return DEFAULT_BUFFER_YEARS + modifier


def _score_entry(
    entry: Dict[str, Any],
    vehicle_lifetime: int,
    threat_level: str,
    pqc_required: bool,
    current_year: int = CURRENT_YEAR,
) -> Dict[str, Any]:
    algo_name = entry.get("recommendedAlgorithm", "Unknown")
    lifespan_str = entry.get("securityLifespan", "")
    dep_year = _extract_deprecation_year(lifespan_str)
    safe_until_year = current_year + vehicle_lifetime
    buffer_years = _compute_buffer_years(threat_level)
    quantum_safe = _is_quantum_safe(entry)

    if pqc_required and not quantum_safe:
        risk = RISK_RED
        reason = (
            f"PQC required, but {algo_name} is quantum-vulnerable. "
            f"Migration: {entry.get('pqcMigrationPath', 'Migrate to PQC algorithm.')}"
        )
    elif quantum_safe:
        risk = RISK_GREEN
        reason = f"{algo_name} is quantum-resistant or symmetric. Lifespan: {lifespan_str}."
    elif dep_year is None:
        risk = RISK_YELLOW
        reason = f"{algo_name} has no explicit deprecation year. Lifespan note: {lifespan_str}."
    elif safe_until_year < (dep_year - buffer_years):
        risk = RISK_GREEN
        reason = (
            f"{algo_name} remains secure through vehicle retirement ({safe_until_year}), "
            f"with buffer before estimated horizon ({dep_year})."
        )
    elif safe_until_year <= dep_year:
        risk = RISK_YELLOW
        reason = (
            f"{algo_name} lifetime window cuts close: vehicle retires in {safe_until_year}, "
            f"estimated horizon is ~{dep_year}. Migration plan recommended."
        )
    else:
        risk = RISK_RED
        reason = (
            f"{algo_name} security horizon ({dep_year}) expires while vehicle is still active "
            f"(until {safe_until_year}). Action required: {entry.get('pqcMigrationPath', 'Plan migration.')}"
        )

    return {
        "algorithm": algo_name,
        "mode": entry.get("mode"),
        "key_length": entry.get("keyLength"),
        "risk": risk,
        "reason": reason,
        "deprecation_year": dep_year,
        "standards": entry.get("standardsRefs", []),
        "migration_path": entry.get("pqcMigrationPath", ""),
        "risk_flags": entry.get("riskFlags", []),
    }


# ---------------------------------------------------------------------------
# PUBLIC API
# ---------------------------------------------------------------------------

def evaluate_use_case(
    params: Dict[str, Any],
    db: Union[Dict[str, Any], List[Dict[str, Any]]]
) -> Dict[str, Any]:
    """
    Filter and score algorithms for a single use case against the knowledge base.
    """
    use_case = params.get("use_case", "")
    vehicle_lifetime = params.get("vehicle_lifetime", 10)
    threat_level = params.get("threat_level", "Medium")
    pqc_required = params.get("pqc_required", False)

    entry = _resolve_use_case_entry(use_case, db)

    if not entry:
        return {
            "use_case": use_case,
            "recommended": [],
            "top_pick": None,
            "error": f"Use case '{use_case}' not found in database."
        }

    scored_item = _score_entry(entry, vehicle_lifetime, threat_level, pqc_required)

    return {
        "use_case": entry.get("useCase", use_case),
        "recommended": [scored_item],
        "top_pick": scored_item["algorithm"],
    }


def generate_batch_summary(recommendations: List[Dict[str, Any]]) -> Dict[str, Any]:
    """
    Roll up multiple evaluate_use_case() results into one combined summary.
    """
    all_entries = []
    for block in recommendations:
        use_case = block.get("use_case", "Unknown")
        for entry in block.get("recommended", []):
            all_entries.append({**entry, "use_case": use_case})

    counts = {"green": 0, "yellow": 0, "red": 0}
    for entry in all_entries:
        if entry["risk"] == RISK_GREEN:
            counts["green"] += 1
        elif entry["risk"] == RISK_YELLOW:
            counts["yellow"] += 1
        elif entry["risk"] == RISK_RED:
            counts["red"] += 1

    if counts["red"] > 0:
        highest_risk = RISK_RED
    elif counts["yellow"] > 0:
        highest_risk = RISK_YELLOW
    else:
        highest_risk = RISK_GREEN

    actionable = [e for e in all_entries if e["risk"] in (RISK_RED, RISK_YELLOW)]
    actionable.sort(key=lambda e: RISK_RANK[e["risk"]])

    priority_actions = []
    for entry in actionable:
        prefix = "URGENT" if entry["risk"] == RISK_RED else "MONITOR"
        dep = f" (~{entry['deprecation_year']})" if entry.get("deprecation_year") else ""
        priority_actions.append(
            f"{prefix}: [{entry['use_case']}] {entry['algorithm']}{dep} — {entry['reason']}"
        )

    return {
        "highest_risk": highest_risk,
        "risk_counts": counts,
        "priority_actions": priority_actions,
    }

File: lib/ruleEngine.js
import fs from 'fs';
import path from 'path';

// ─── Constants ───────────────────────────────────────────────────────────────
const CURRENT_YEAR = 2026;
const DEFAULT_BUFFER_YEARS = 4;

const THREAT_BUFFER_MODIFIER = {
  Low: -1,
  Medium: 0,
  High: 1,
  Critical: 2,
};

const RISK_GREEN = '🟢';
const RISK_YELLOW = '🟡';
const RISK_RED = '🔴';
const RISK_RANK = { [RISK_GREEN]: 0, [RISK_YELLOW]: 1, [RISK_RED]: 2 };

// Synonym mapping for fuzzy use-case resolution
const SYNONYMS = {
  otaupdates: 'Signatures',
  ota: 'Signatures',
  otaupdate: 'Signatures',
  firmwaresigning: 'Signatures',
  firmware: 'Signatures',
  codesigning: 'Signatures',
  v2x: 'Communication',
  v2xcommunication: 'Communication',
  vehicletovehicle: 'Communication',
  v2v: 'Communication',
  v2i: 'Communication',
  can: 'MAC_SecOC',
  canbus: 'MAC_SecOC',
  canfd: 'MAC_SecOC',
  secoc: 'MAC_SecOC',
  messageauthentication: 'MAC_SecOC',
  invehicle: 'MAC_SecOC',
  storage: 'DataEncryption',
  encryption: 'DataEncryption',
  dataatrest: 'DataEncryption',
  confidentiality: 'DataEncryption',
  boot: 'SecureBoot',
  bootloader: 'SecureBoot',
  bootchain: 'SecureBoot',
  lifecycle: 'KeyManagement',
  keyrotation: 'KeyManagement',
  keystorage: 'KeyManagement',
  provisioning: 'HSMProvisioning',
  keyinjection: 'HSMProvisioning',
  factory: 'HSMProvisioning',
  hash: 'Integrity',
  checksum: 'Integrity',
  sha: 'Integrity',
  random: 'RNG',
  entropy: 'RNG',
  seed: 'RNG',
  digitalsignatures: 'Signatures',
  digitalsignature: 'Signatures',
  signatures: 'Signatures',
  keyexchange: 'KeyExchange',
  keyagreement: 'KeyExchange',
  sessionkey: 'KeyExchange',
  handshake: 'KeyExchange',
  rng: 'RNG',
  randomnumber: 'RNG',
  integrity: 'Integrity',
  secureboot: 'SecureBoot',
  dataencryption: 'DataEncryption',
  macsecoc: 'MAC_SecOC',
  keymanagement: 'KeyManagement',
  hsmprovisioning: 'HSMProvisioning',
};

// ─── Helpers ─────────────────────────────────────────────────────────────────

function loadKnowledgeBase() {
  const filePath = path.join(process.cwd(), 'data', 'knowledge-base.json');
  const raw = fs.readFileSync(filePath, 'utf-8');
  return JSON.parse(raw);
}

/** Extract the first four-digit year (20xx) from a lifespan description. */
function extractDeprecationYear(securityLifespan) {
  if (!securityLifespan) return null;
  const matches = securityLifespan.match(/\b(20\d\d)\b/g);
  return matches ? parseInt(matches[0], 10) : null;
}

/** Heuristic: is the recommended algorithm post-quantum safe? */
function isQuantumSafe(entry) {
  const text = `${entry.pqcMigrationPath || ''} ${entry.rationale || ''} ${entry.recommendedAlgorithm || ''}`.toLowerCase();
  if (text.includes('no migration needed') || text.includes('not quantum-vulnerable') || text.includes('not applicable')) return true;
  if (text.includes('quantum-resistant') || text.includes('quantum resistant')) return true;
  if ((entry.recommendedAlgorithm || '').toLowerCase().includes('crystals')) return true;
  return false;
}

/** Strip non-alphanumeric and lowercase for fuzzy matching. */
function normalizeKey(key) {
  return String(key).replace(/[^a-zA-Z0-9]/g, '').toLowerCase();
}

/** Resolve a user-supplied use-case string to a knowledge-base entry. */
function resolveUseCaseEntry(useCase, kb) {
  if (!useCase) return null;
  const normTarget = normalizeKey(useCase);

  // Direct key match
  for (const [k, v] of Object.entries(kb)) {
    if (normalizeKey(k) === normTarget || normalizeKey(v.useCase || '') === normTarget) {
      return { entry: v, resolvedKey: k };
    }
  }

  // Synonym match
  const mappedKey = SYNONYMS[normTarget];
  if (mappedKey && kb[mappedKey]) return { entry: kb[mappedKey], resolvedKey: mappedKey };

  return null;
}

// ─── Risk Scoring ────────────────────────────────────────────────────────────

function scoreEntry(entry, vehicleLifetimeYears, threatLevel, pqcRequired) {
  const algoName = entry.recommendedAlgorithm || 'Unknown';
  const lifespanStr = entry.securityLifespan || '';
  const depYear = extractDeprecationYear(lifespanStr);
  const safeUntilYear = CURRENT_YEAR + vehicleLifetimeYears;
  const bufferYears = DEFAULT_BUFFER_YEARS + (THREAT_BUFFER_MODIFIER[threatLevel] || 0);
  const quantumSafe = isQuantumSafe(entry);

  let risk, reason;

  if (pqcRequired && !quantumSafe) {
    risk = RISK_RED;
    reason = `PQC required but ${algoName} is quantum-vulnerable. ${entry.pqcMigrationPath || 'Migrate to a PQC algorithm.'}`;
  } else if (quantumSafe) {
    risk = RISK_GREEN;
    reason = `${algoName} is quantum-resistant or symmetric — ${lifespanStr}.`;
  } else if (depYear === null) {
    risk = RISK_YELLOW;
    reason = `${algoName} has no explicit deprecation year. Note: ${lifespanStr}.`;
  } else if (safeUntilYear < depYear - bufferYears) {
    risk = RISK_GREEN;
    reason = `${algoName} remains secure through vehicle retirement (${safeUntilYear}), well within estimated horizon (~${depYear}).`;
  } else if (safeUntilYear <= depYear) {
    risk = RISK_YELLOW;
    reason = `${algoName} lifetime cuts close — vehicle retires ${safeUntilYear}, horizon ~${depYear}. Plan migration.`;
  } else {
    risk = RISK_RED;
    reason = `${algoName} horizon (~${depYear}) expires while vehicle is still active (until ${safeUntilYear}). ${entry.pqcMigrationPath || 'Immediate action required.'}`;
  }

  return {
    algorithm: algoName,
    mode: entry.mode,
    keyLength: entry.keyLength,
    risk,
    reason,
    deprecationYear: depYear,
    standards: entry.standardsRefs || [],
    migrationPath: entry.pqcMigrationPath || '',
    riskFlags: entry.riskFlags || [],
    rationale: entry.rationale || '',
    securityLifespan: lifespanStr,
  };
}

// ─── Public API ──────────────────────────────────────────────────────────────

/**
 * Get a recommendation for a single use case.
 */
export function getRecommendation(useCaseKey, vehicleLifetimeYears = 15, threatLevel = 'Medium', pqcRequired = false) {
  const kb = loadKnowledgeBase();
  const resolved = resolveUseCaseEntry(useCaseKey, kb);

  if (!resolved) {
    return {
      error: `Unknown use case: "${useCaseKey}"`,
      useCase: useCaseKey,
      recommended: [],
    };
  }

  const { entry } = resolved;
  const scored = scoreEntry(entry, vehicleLifetimeYears, threatLevel, pqcRequired);

  // PQC urgency label
  let pqcUrgency;
  if (vehicleLifetimeYears >= 15) pqcUrgency = 'High — plan PQC migration now';
  else if (vehicleLifetimeYears >= 10) pqcUrgency = 'Medium — monitor PQC transition timeline';
  else pqcUrgency = 'Low — current algorithm adequate for this lifetime';

  return {
    useCase: entry.useCase,
    ...scored,
    pqcUrgency,
  };
}

/**
 * Get recommendations for multiple use cases with a combined summary.
 */
export function getMultipleRecommendations(queries) {
  const results = queries.map((q) =>
    getRecommendation(
      q.useCaseKey || q.use_case || '',
      Number(q.vehicleLifetimeYears || q.vehicle_lifetime) || 15,
      q.threatLevel || q.threat_level || 'Medium',
      Boolean(q.pqcRequired || q.pqc_required),
    ),
  );

  const valid = results.filter((r) => !r.error);

  // Risk counts
  const counts = { green: 0, yellow: 0, red: 0 };
  for (const r of valid) {
    if (r.risk === RISK_GREEN) counts.green++;
    else if (r.risk === RISK_YELLOW) counts.yellow++;
    else if (r.risk === RISK_RED) counts.red++;
  }

  // Highest risk
  let highestRisk = RISK_GREEN;
  if (counts.red > 0) highestRisk = RISK_RED;
  else if (counts.yellow > 0) highestRisk = RISK_YELLOW;

  // Priority actions
  const actionable = valid
    .filter((r) => r.risk === RISK_RED || r.risk === RISK_YELLOW)
    .sort((a, b) => (RISK_RANK[b.risk] || 0) - (RISK_RANK[a.risk] || 0));

  const priorityActions = actionable.map((r) => {
    const prefix = r.risk === RISK_RED ? 'URGENT' : 'MONITOR';
    const dep = r.deprecationYear ? ` (~${r.deprecationYear})` : '';
    return `${prefix}: [${r.useCase}] ${r.algorithm}${dep} — ${r.reason}`;
  });

  // Most critical
  const sorted = [...valid].sort((a, b) => (RISK_RANK[b.risk] || 0) - (RISK_RANK[a.risk] || 0));
  const mostCritical = sorted[0];

  return {
    individual: results,
    combinedSummary: {
      highestRisk,
      riskCounts: counts,
      priorityActions,
      mostCriticalUseCase: mostCritical?.useCase || null,
    },
  };
}

/**
 * Return list of all available use cases from the knowledge base.
 */
export function getAvailableUseCases() {
  const kb = loadKnowledgeBase();
  return Object.entries(kb).map(([key, value]) => ({
    key,
    label: value.useCase,
    mode: value.mode,
  }));
}